modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotFullCheckDistributedProcess.java

```diff
@@ -0,0 +1,549 @@
+import org.apache.ignite.events.DiscoveryEvent;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.cluster.ClusterTopologyCheckedException;
+import org.apache.ignite.internal.management.cache.IdleVerifyResultV2;
+import org.apache.ignite.internal.management.cache.PartitionKeyV2;
+import org.apache.ignite.internal.management.cache.VerifyBackupPartitionsTaskV2;
+import org.apache.ignite.internal.processors.cache.verify.PartitionHashRecordV2;
+import org.apache.ignite.internal.util.distributed.DistributedProcess;
+import org.apache.ignite.internal.util.future.GridFinishedFuture;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.events.EventType.EVT_NODE_FAILED;
+import static org.apache.ignite.events.EventType.EVT_NODE_LEFT;
+import static org.apache.ignite.internal.util.distributed.DistributedProcess.DistributedProcessType.SNAPSHOT_CHECK_METAS;
+import static org.apache.ignite.internal.util.distributed.DistributedProcess.DistributedProcessType.SNAPSHOT_VALIDATE_PARTS;
+
+/** Distributed process of snapshot full checking (with the partition hashes). */
+public class SnapshotFullCheckDistributedProcess {
+    /** */
+    private static final IgniteInternalFuture FINISHED_FUT = new GridFinishedFuture<>();
+
+    /** */
+    private final IgniteLogger log;
+
+    /** */
+    private final GridKernalContext kctx;
+
+    /** Snapshot check requests per snapshot on every node. */
+    final Map<String, SnapshotFullCheckOperationRequest> requests = new ConcurrentHashMap<>();
+
+    /** Cluster-wide operation futures per snapshot called from current node. */
+    private final Map<UUID, GridFutureAdapter<SnapshotPartitionsVerifyTaskResult>> clusterOpFuts = new ConcurrentHashMap<>();
+
+    /** Check metas first phase subprocess. */
+    private final DistributedProcess<SnapshotFullCheckOperationRequest, ArrayList<SnapshotMetadata>> phase1CheckMetas;
+
+    /** Partition hashes second phase subprocess.  */
+    private final DistributedProcess<SnapshotFullCheckOperationRequest, HashMap<PartitionKeyV2, PartitionHashRecordV2>> phase2PartsHashes;
+
+    /** */
+    public SnapshotFullCheckDistributedProcess(GridKernalContext kctx) {
+        this.kctx = kctx;
+
+        log = kctx.log(getClass());
+
+        phase1CheckMetas = new DistributedProcess<>(kctx, SNAPSHOT_CHECK_METAS, this::prepareAndCheckMetas,
+            this::reducePreparationAndMetasCheck);
+
+        phase2PartsHashes = new DistributedProcess<>(kctx, SNAPSHOT_VALIDATE_PARTS, this::validateParts,
+            this::reduceValidatePartsAndFinish);
+
+        kctx.event().addLocalEventListener((evt) -> nodeLeft(((DiscoveryEvent)evt).eventNode()), EVT_NODE_FAILED, EVT_NODE_LEFT);
```

Comment:
Yes. But I wanted to make the code more independent. The process is able to handle most of the events itself. WDYT?

Reply:
I don't think that a code duplication definitely means a problem. IMHO, much worse is a violation of the principles like the encapsulation. SnapshotManager ignores clients:

if (ctx.clientNode())
    return;
            
It won't deliver events to the check process in some cases. Why? Why it should decide what to do with an event instead of the dedicated process? Client can start snapshot checking too. It keeps the launch future, results and partial results. It requires cleaning too. Several tests fail on it. Client must know about a node leave. But IgniteSnasphot manager doesn't allow. As I remember, I fixed a couple of similar bugs in the snapshots: we ignored client and non-baseline node.